File: viewer/src/main/scala/farjs/viewer/quickview/QuickViewPlugin.scala
package farjs.viewer.quickview

import farjs.filelist.FileListPlugin
import farjs.filelist.stack.{PanelStackItem, PanelStacks}
import scommons.react.ReactClass

import scala.scalajs.js

object QuickViewPlugin extends FileListPlugin(js.Array("C-q")) {

  private val panelComp: ReactClass = QuickViewPanel()

  override def onKeyTrigger(key: String,
                            stacks: PanelStacks,
                            data: js.UndefOr[js.Dynamic] = js.undefined): js.Promise[js.UndefOr[ReactClass]] = {
    val exists =
      if (stacks.left.stack.peek().component == panelComp) {
        stacks.left.stack.pop()
        true
      }
      else if (stacks.right.stack.peek().component == panelComp) {
        stacks.right.stack.pop()
        true
      }
      else false
    
    if (!exists) {
      val stack = PanelStacks.nonActive(stacks).stack
      stack.push(PanelStackItem[QuickViewParams](
        component = panelComp,
        dispatch = js.undefined,
        actions = js.undefined,
        state = QuickViewParams()
      ))
    }
    
    js.Promise.resolve[js.UndefOr[ReactClass]](js.undefined)
  }
}

File: viewer/quickview/QuickViewPlugin.mjs
/**
 * @typedef {import("@farjs/filelist/stack/WithStacksProps.mjs").WithStacksProps} WithStacksProps
 * @import { ReactComponent } from "@farjs/filelist/FileListData.mjs"
 * @import { QuickViewParams } from "./QuickViewDir.mjs";
 */
import FileListPlugin from "@farjs/filelist/FileListPlugin.mjs";
import WithStacksProps from "@farjs/filelist/stack/WithStacksProps.mjs";
import PanelStackItem from "@farjs/filelist/stack/PanelStackItem.mjs";
import QuickViewPanel from "./QuickViewPanel.mjs";

class QuickViewPluginImpl extends FileListPlugin {
  constructor() {
    super(["C-q"]);

    /** @type {ReactComponent} */
    this.panelComp = QuickViewPanel;
  }

  /**
   * @param {string} _
   * @param {WithStacksProps} stacks
   * @returns {Promise<ReactComponent | undefined>}
   */
  onKeyTrigger(_, stacks) {
    const exists = (() => {
      if (stacks.left.stack.peek().component === this.panelComp) {
        stacks.left.stack.pop();
        return true;
      }
      if (stacks.right.stack.peek().component === this.panelComp) {
        stacks.right.stack.pop();
        return true;
      }

      return false;
    })();

    if (!exists) {
      /** @type {QuickViewParams} */
      const params = {
        name: "",
        parent: "",
        folders: 0,
        files: 0,
        filesSize: 0,
      };

      const stack = WithStacksProps.nonActive(stacks).stack;
      stack.push(
        new PanelStackItem(this.panelComp, undefined, undefined, params)
      );
    }

    return Promise.resolve(undefined);
  }
}

const QuickViewPlugin = new QuickViewPluginImpl();

export default QuickViewPlugin;
